File: node/src/merchant-address.ts
import { HttpClient } from './http-client';
import type { ShipzyConfig } from './http-client';
import type { ApiResult } from './index';

// ============ Types ============

export interface TenantAddress {
    id: string;
    full_name?: string;
    company_name?: string;
    street?: string;
    house_number?: string;
    postal_code?: string;
    city?: string;
    country_code?: string;
    phone?: string;
    email?: string;
    is_default?: boolean;
    created_at: string;
    updated_at: string;
}

export interface TenantAddressListResponse {
    data: TenantAddress[];
    total: number;
    page: number;
    page_size: number;
}

// ============ Client ============

export class MerchantAddressClient extends HttpClient {
    async list(filter: { page?: number; page_size?: number; role_tag?: string } = {}): Promise<ApiResult<TenantAddressListResponse>> {
        return this.request('/api/v1/merchant/addresses/list', 'POST', filter);
    }

    async create(body: Partial<TenantAddress>): Promise<ApiResult<TenantAddress>> {
        return this.request('/api/v1/merchant/addresses/create', 'POST', body);
    }

    async update(id: string, updates: Partial<TenantAddress>): Promise<ApiResult<{ id: string }>> {
        return this.request(`/api/v1/merchant/addresses/${id}/update`, 'POST', updates);
    }

    async delete(id: string): Promise<ApiResult<{ id: string }>> {
        return this.request(`/api/v1/merchant/addresses/${id}/delete`, 'POST', {});
    }

    async setDefault(id: string, type: 'sender' | 'return' | 'contact' | 'warehouse'): Promise<ApiResult<{ id: string; type: string }>> {
        return this.request(`/api/v1/merchant/addresses/${id}/set-default`, 'POST', { type });
    }
}

File: rust/src/merchant_address.rs
use reqwest::Method;

use crate::error::Result;
use crate::http_client::HttpClient;
use crate::types::{AddressItem, ListResponse};

pub struct MerchantAddressClient {
    pub(crate) inner: HttpClient,
}

impl MerchantAddressClient {
    pub fn new(inner: HttpClient) -> Self {
        Self { inner }
    }
}

impl MerchantAddressClient {
    pub async fn list(
        &self,
        filter: Option<serde_json::Value>,
    ) -> Result<ListResponse<AddressItem>> {
        self.inner
            .request(
                "/api/v1/merchant/addresses/list",
                Method::POST,
                filter,
            )
            .await
    }

    pub async fn get(&self, id: &str) -> Result<AddressItem> {
        let resp: ListResponse<AddressItem> = self
            .inner
            .request(
                "/api/v1/merchant/addresses/list",
                Method::POST,
                Some(serde_json::json!({ "id": id })),
            )
            .await?;
        resp.data
            .into_iter()
            .find(|a| a.id == id)
            .ok_or_else(|| crate::error::ShipzyError::Http {
                status: 404,
                message: format!("Address {} not found", id),
            })
    }

    pub async fn create(&self, body: serde_json::Value) -> Result<AddressItem> {
        self.inner
            .request(
                "/api/v1/merchant/addresses/create",
                Method::POST,
                Some(body),
            )
            .await
    }

    pub async fn update(
        &self,
        id: &str,
        body: serde_json::Value,
    ) -> Result<AddressItem> {
        self.inner
            .request(
                &format!("/api/v1/merchant/addresses/{}/update", id),
                Method::POST,
                Some(body),
            )
            .await
    }

    pub async fn delete(&self, id: &str) -> Result<serde_json::Value> {
        self.inner
            .request(
                &format!("/api/v1/merchant/addresses/{}/delete", id),
                Method::POST,
                None,
            )
            .await
    }

    pub async fn set_default(
        &self,
        id: &str,
        address_type: &str,
    ) -> Result<serde_json::Value> {
        self.inner
            .request(
                &format!("/api/v1/merchant/addresses/{}/set-default", id),
                Method::POST,
                Some(serde_json::json!({ "type": address_type })),
            )
            .await
    }
}
